### PR TITLE
Add float SIG_VALTYPE_ types to DBC string dump

### PR DESCRIPTION
**Pull request by karlding**
Currently the `as_dbc_string` function ignores `float` and `double` signal
attributes that identify a `Signal` as a floating type. As such, when a
DBC containing floating types is loaded and subsequently dumped, we lose
information regarding these floating-point `Signals`.

This fixes `as_dbc_string` to also dump `SIG_VALTYPE_`.

### CODE REVIEW

File: tests/test_database.py [hunk 1]
```diff
@@ -4029,6 +4030,26 @@ def test_float_dbc(self):
...
+        message = new_db.get_message_by_frame_id(1025)
+        signal = message.get_signal_by_name('Signal1')
+        self.assertEqual(signal.is_float, True)
+        self.assertEqual(signal.length, 32)
+        signal = message.get_signal_by_name('Signal2')
+        self.assertEqual(signal.is_float, True)
+        self.assertEqual(signal.length, 32)
+
+        # Assert DBC contents
+        self.assertIn('SIG_VALTYPE_ 1024 Signal1 : 2;', db.as_dbc_string())
```
Comment (karlding) [inline tests/test_database.py]: I noticed that `as_dbc_string` dumps extra fields that are not present in all the DBC files.

```
	BO_TX_BU_
	BA_DEF_REL_
	BA_REL_
	BA_DEF_DEF_REL_
	BU_SG_REL_
	BU_EV_REL_
	BU_BO_REL_
	SG_MUL_VAL_
```

Do we expect that loading and storing a DBC is invariant, given a well-formed DBC file? Or are these actually needed in all cases, and the files that are missing these do not conform to the DBC standard?

Comment (eerimoq) [reply tests/test_database.py]: There is no DBC standard, is there?

Those fields originally was copied from a random DBC-file, and more fields were probably added later to make CANdb++ able to load the dumped files.

Comment (qc17) [reply tests/test_database.py]: Yes, there is no official standard but from vector's  "DBC File Format Documentation":

Signals with value types 'float' and 'double' have additional entries in the signal_
valtype_list section.

`signal_extended_value_type_list = 'SIG_VALTYPE_' message_id signal_
name signal_extended_value_type ';' ;`

`signal_extended_value_type = '0' | '1' | '2' | '3' ; (* 0=signed or
unsigned integer, 1=32-bit IEEE-float, 2=64-bit IEEE-double *)`

Comment (karlding) [reply tests/test_database.py]: Cool, I was only wondering because otherwise I was going to just add those to the DBCs missing them and then do a direct diff of the files.

Also, if I understand that grammar correctly, it implies that extended value types should always be there, since they enumerate all 4 possible cases, but all of the DBCs without floating types in the repo seem to be missing this. Should those fields be added to those DBCs?

Comment (eerimoq) [reply tests/test_database.py]: Sounds like a good plan to add these fields to the ``NS_`` section in all DBC-files, as long as CANdb++ can load them and that it isn't in conflict with the DBC File Format Documentation.

Comment (karlding) [reply tests/test_database.py]: I don't have access to CANdb++, so I won't be able to verify that. Do either of you?

Comment (qc17) [reply tests/test_database.py]: `SIG_VALTYPE_` always appears in the 'New Symbol' section although actual optional:

`version = ['VERSION' '"' { CANdb_version_string } '"' ];`

`new_symbols = [ 'NS_' ':' ['CM_'] ['BA_DEF_'] ['BA_'] ['VAL_']
['CAT_DEF_'] ['CAT_'] ['FILTER'] ['BA_DEF_DEF_'] ['EV_DATA_']
['ENVVAR_DATA_'] ['SGTYPE_'] ['SGTYPE_VAL_'] ['BA_DEF_SGTYPE_']
['BA_SGTYPE_'] ['SIG_TYPE_REF_'] ['VAL_TABLE_'] ['SIG_GROUP_']
['SIG_VALTYPE_'] ['SIGTYPE_VALTYPE_'] ['BO_TX_BU_']
['BA_DEF_REL_'] ['BA_REL_'] ['BA_DEF_DEF_REL_'] ['BU_SG_REL_']
['BU_EV_REL_'] ['BU_BO_REL_'] ];`

u32-Test:

---8<---------
VERSION "None"

NS_ : 
	NS_DESC_
	CM_
	BA_DEF_
	BA_
	VAL_
	CAT_DEF_
	CAT_
	FILTER
	BA_DEF_DEF_
	EV_DATA_
	ENVVAR_DATA_
	SGTYPE_
	SGTYPE_VAL_
	BA_DEF_SGTYPE_
	BA_SGTYPE_
	SIG_TYPE_REF_
	VAL_TABLE_
	SIG_GROUP_
	SIG_VALTYPE_
	SIGTYPE_VALTYPE_
	BO_TX_BU_
	BA_DEF_REL_
	BA_REL_
	BA_DEF_DEF_REL_
	BU_SG_REL_
	BU_EV_REL_
	BU_BO_REL_
	SG_MUL_VAL_

BS_:

BU_: AuSy INCA

BO_ 1 AuSy_MSG_ID1: 8 AuSy
 SG_ u32_2 : 32|1@1+ (1,0) [0|0] ""  INCA
 SG_ u32_1 : 0|1@1+ (1,0) [0|0] ""  INCA

CM_ SG_ 1 u32_2 "bar";
CM_ SG_ 1 u32_1 "foo";
---8<---------

I habe a DBCdb++ available for some tests.

Comment (karlding) [reply tests/test_database.py]: Perhaps I should clarify what I was asking.

I was referring to your grammar that you provided:

```
signal_extended_value_type_list = 'SIG_VALTYPE_' message_id signal_name signal_extended_value_type ';' ;

signal_extended_value_type = '0' | '1' | '2' | '3' ; (* 0=signed or unsigned integer, 1=32-bit IEEE-float, 2=64-bit IEEE-double *)
```

Does `SIG_VALTYPE_ message_id signal_name 0;` ever show up? According to this grammar, it seems to imply that it does, but I haven't seen it in any DBCs.

The second question was whether these fields were actually always present in the DBCs:

```
	BO_TX_BU_
	BA_DEF_REL_
	BA_REL_
	BA_DEF_DEF_REL_
	BU_SG_REL_
	BU_EV_REL_
	BU_BO_REL_
	SG_MUL_VAL_
```

Currently cantools always adds `SIG_VALTYPE_` in the New Symbol section, like you point out CANdb++ does. If CANdb++ does not include these fields, I was proposing:

1. Removing them from cantools' DBC string if not needed, and then updating all the DBC test files that extraneously include them by re-saving them in CANdb++
1. Always dumping these optional fields in the DBC string for simplicity if CANdb++ can open them (and then updating DBC test files missing these fields to always contain them) and nobody complains about it

Basically I was wondering whether CANdb++ would strip extraneous symbols if they weren't necessary, and whether cantools should do the same.

Comment (eerimoq) [reply tests/test_database.py]: Let's only dump `SIG_VALTYPE_` for floating point numbers.

Let's leave the `NS_` section as is, and potentially make it match the CANdb++ output later.

Comment (qc17) [reply tests/test_database.py]: As far as I know CANdb++ only writes `signal_extended_value_type = '1' | '2'  ;`
Empty symbols are not needed in the `NS_`-section but will be appended by CANdb++.
